# HG changeset patch
# User Michael Goffioul <michael.goffioul@gmail.com>
# Date 1377272824 14400
#      Fri Aug 23 11:47:04 2013 -0400
# Node ID ea23eb07f8ed2ac02bdb3f112d1c3e1c0da36421
# Parent  27fc61cfbc6eaf4a9f9fa0f05cac6096947baa0b
Add missing TeX symbols and fix few others (bug #39828).

* libinterp/corefcn/oct-tex-lexer.ll (ID): Allow digit characters.
* libinterp/corefcn/txt-eng.cc (symbol_names): Add "angle", "ast",
"sim", "Leftarrow" and "Rightarrow".
(symbol_codes): Likewise. Fix unicode values for symbols "langle" and
"rangle".

diff --git a/libinterp/corefcn/oct-tex-lexer.ll b/libinterp/corefcn/oct-tex-lexer.ll
--- a/libinterp/corefcn/oct-tex-lexer.ll
+++ b/libinterp/corefcn/oct-tex-lexer.ll
@@ -34,17 +34,17 @@ along with Octave; see the file COPYING.
 #include "oct-tex-parser.h"
 }
 
 %x	NUM_MODE
 %x	MAYBE_NUM_MODE
 %x	COMMAND
 
 D       [0-9]
-ID	[a-zA-Z]
+ID	[a-zA-Z0-9]
 NUM	(({D}+\.?{D}*)|(\.{D}+))
 
 %%
 
 %{
 // Numeric values
 %}
 
diff --git a/libinterp/corefcn/txt-eng.cc b/libinterp/corefcn/txt-eng.cc
--- a/libinterp/corefcn/txt-eng.cc
+++ b/libinterp/corefcn/txt-eng.cc
@@ -23,16 +23,18 @@ along with Octave; see the file COPYING.
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
 #include "txt-eng.h"
 
 static const char* symbol_names[] = {
     "alpha",
+    "angle",
+    "ast",
     "beta",
     "gamma",
     "delta",
     "epsilon",
     "zeta",
     "eta",
     "theta",
     "vartheta",
@@ -90,26 +92,29 @@ static const char* symbol_names[] = {
     "lceil",
     "cdot",
     "neg",
     "times",
     "surd",
     "varpi",
     "rangle",
 
+    "sim",
     "leq",
     "infty",
     "clubsuit",
     "diamondsuit",
     "heartsuit",
     "spadesuit",
     "leftrightarrow",
     "leftarrow",
+    "Leftarrow",
     "uparrow",
     "rightarrow",
+    "Rightarrow",
     "downarrow",
     "circ",
     "pm",
     "geq",
     "propto",
     "partial",
     "bullet",
     "div",
@@ -131,16 +136,18 @@ static const char* symbol_names[] = {
 };
 
 // Maps the symbol names (using index from symbol_names array) to
 // character codes, using 2 mapping:
 // - Unicode
 // - MS symbol (using Private Use Area)
 static uint32_t symbol_codes[][2] = {
       { 0x03B1, 0xF061 },		// alpha
+      { 0x2220, 0xF0D0 },               // angle
+      { 0x2217, 0xF02A },               // ast
       { 0x03B2, 0xF062 },		// beta
       { 0x03B3, 0xF067 },		// gamma
       { 0x03B4, 0xF064 },		// delta
       { 0x03B5, 0xF065 },		// epsilon
       { 0x03B6, 0xF07A },		// zeta
       { 0x03B7, 0xF068 },		// eta
       { 0x03B8, 0xF071 },		// theta
       { 0x03D1, 0xF04A },		// vartheta
@@ -162,17 +169,17 @@ static uint32_t symbol_codes[][2] = {
       { 0x2283, 0xF0C9 },		// supset
       { 0x222B, 0xF0F2 },		// int
       { 0x230B, 0xF0FB },		// rfloor
       { 0x230A, 0xF0EB },		// lfloor
       { 0x22A5, 0xF05E },		// perp
       { 0x2227, 0xF0D9 },		// wedge
       { 0x2309, 0xF0F9 },		// rceil
       { 0x2228, 0xF0DA },		// vee
-      { 0x2220, 0xF0E1 },		// langle
+      { 0x27E8, 0xF0E1 },		// langle
 
       { 0x03C5, 0xF075 },		// upsilon
       { 0x03C6, 0xF066 },		// phi
       { 0x03C7, 0xF063 },		// chi
       { 0x03C8, 0xF079 },		// psi
       { 0x03C9, 0xF077 },		// omega
       { 0x0393, 0xF047 },		// Gamma
       { 0x0394, 0xF044 },		// Delta
@@ -196,28 +203,31 @@ static uint32_t symbol_codes[][2] = {
       { 0x2286, 0xF0CD },		// subseteq
       { 0x2208, 0xF0CE },		// in
       { 0x2308, 0xF0E9 },		// lceil
       { 0x22C5, 0xF0D7 },		// cdot
       { 0x00AC, 0xF0D8 },		// neg
       { 0x00D7, 0xF0B4 },		// times
       { 0x221A, 0xF0D6 },		// surd
       { 0x03D6, 0xF076 },		// varpi
-      { 0x232A, 0xF0F1 },		// rangle
+      { 0x27E9, 0xF0F1 },		// rangle
 
+      { 0x223C, 0xF07E },               // sim
       { 0x2264, 0xF0A3 },		// leq
       { 0x221E, 0xF0A5 },		// infty
       { 0x2663, 0xF0A7 },		// clubsuit
       { 0x2666, 0xF0A8 },		// diamondsuit
       { 0x2665, 0xF0A9 },		// heartsuit
       { 0x2660, 0xF0AA },		// spadesuit
       { 0x2194, 0xF0AB },		// leftrightarrow
       { 0x2190, 0xF0AC },		// leftarrow
+      { 0x21D0, 0xF0DC },		// Leftarrow
       { 0x2191, 0xF0AD },		// uparrow
       { 0x2192, 0xF0AE },		// rightarrow
+      { 0x21D2, 0xF0DE },		// Rightarrow
       { 0x2193, 0xF0AF },		// downarrow
       { 0x25CB, 0xF0B0 },		// circ
       { 0x00B1, 0xF0B1 },		// pm
       { 0x2265, 0xF0B3 },		// geq
       { 0x221D, 0xF0B5 },		// propto
       { 0x2202, 0xF0B6 },		// partial
       { 0x2022, 0xF0B7 },		// bullet
       { 0x00F7, 0xF0B8 },		// div
